validator_client/slashing_protection/src/slashing_protection.rs

```diff
@@ -0,0 +1,378 @@
+    /// If the error returned is an IOError, do not sign nor broadcast the attestation.
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe>;
+}
+
+impl SafeFromSlashing<SignedAttestation> for HistoryInfo<SignedAttestation> {
+    type U = AttestationData;
+
+    fn verify_and_get_index(
+        &self,
+        incoming_data: &AttestationData,
+        data_history: &[SignedAttestation],
+    ) -> Result<Safe, NotSafe> {
+        check_for_attester_slashing(incoming_data, data_history)
+    }
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
+        let check = self.verify_and_get_index(incoming_data, &self.data[..]);
+        match check {
+            Ok(safe) => match safe.reason {
+                ValidityReason::SameVote => Ok(()),
+                _ => self
+                    .insert_and_write(SignedAttestation::from(incoming_data), safe.insert_index)
+                    .map_err(|e| NotSafe::IOError(e.kind())),
+            },
+            Err(notsafe) => Err(notsafe),
+        }
+    }
+}
+
+impl SafeFromSlashing<SignedBlock> for HistoryInfo<SignedBlock> {
+    type U = BeaconBlockHeader;
+
+    fn verify_and_get_index(
+        &self,
+        incoming_data: &BeaconBlockHeader,
+        data_history: &[SignedBlock],
+    ) -> Result<Safe, NotSafe> {
+        check_for_proposer_slashing(incoming_data, data_history)
+    }
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
+        let check = self.verify_and_get_index(incoming_data, &self.data[..]);
+        match check {
+            Ok(safe) => match safe.reason {
+                // Casting the same vote, no need to add it go the history
+                ValidityReason::SameVote => Ok(()),
+                // New attestation, add it to memory and file history
+                _ => self
+                    .insert_and_write(SignedBlock::from(incoming_data), safe.insert_index)
+                    .map_err(|e| NotSafe::IOError(e.kind())),
+            },
+            Err(notsafe) => Err(notsafe),
+        }
+    }
+}
+
+/// Struct used for checking if attestations or blockheader are safe from slashing.
+#[derive(Debug)]
+pub struct HistoryInfo<T: Encode + Decode + Clone + PartialEq> {
+    filepath: PathBuf,
+    data: Vec<T>,
+}
+
+impl<T: Encode + Decode + Clone + PartialEq> PartialEq for HistoryInfo<T> {
+    fn eq(&self, other: &Self) -> bool {
+        let my_data = &self.data;
+        let other_data = &other.data;
+        self.filepath == other.filepath && *my_data == *other_data
+    }
+}
+
+impl<T: Encode + Decode + Clone + PartialEq> HistoryInfo<T> {
+    /// Inserts the incoming data in the in-memory history, and writes it to the history file.
+    pub fn insert_and_write(&mut self, data: T, index: usize) -> IOResult<()> {
+        self.data.insert(index, data);
```

Comment:
should we `assert!(index < data.len())` ?

Reply:
Since this is a public function, I think we should return an error if it will cause a panic on `insert`.

However, perhaps this should be a private function?